### ISSUE TITLE
JSON Formatted Output Support

### ISSUE THREAD

**Issue by RichiCoder1** [microsoft/winget-cli#184]
# Description of the new feature/enhancement

To enable automation and consumption via other tools, it'd be great if current and future commands supported a JSON output mode.

**Comment by mikemaccana** [comment]
Modern-day Windows command line apps output structured objects. These can already be converted to JSON, Excel, CSV, or whatever other format is preferred. The issue is that winget doesn't output any structured data. See #196

**Comment by RichiCoder1** [comment]
@mikemaccana As mentioned on the other two issues, PowerShell isn't available in all Windows SKUs which this is targeting. For native PowerShell support, you're looking for https://github.com/microsoft/winget-cli/issues/221.

**Comment by megamorf** [comment]
@RichiCoder1 Could you list all target SKUs and/or a link to the requirements spec, please?

**Comment by RichiCoder1** [comment]
It's been mentioned multiple times across multiple sources:
* https://github.com/microsoft/winget-cli/issues/221#issuecomment-632985508
* https://twitter.com/richturn_ms/status/1262973495559876608?s=20

**Comment by denelon** [comment]
@RichiCoder1 I'm going through the backlog as a part of our planning exercise for the next few milestones. I'm going to close this as a Duplicate of #221.

**Comment by RichiCoder1** [comment]
I'm cool with that as long as the ultimate functionality is still flexible enough to be consumed from other tools

**Comment by denelon** [comment]
Yes, that is the intent. Rather than building some kind of `--output-JSON` flag, the native PowerShell implementation will provide objects that may be used, and will not have the truncated output string behavior currently exhibited in the winget.exe client.

**Comment by mikemaccana** [comment]
Woo native powershell implementation! That’s exactly what I personally
wanted.

**Comment by denelon** [comment]
I'm going to re-open this issue. It's not high-priority (unless we see a lot of 👍) since we have the option in PowerShell, but I can see value in what is being asked for.

We still have more work to completely move everything the CLI is calling into the COM APIs, and once that's done, providing some kind of output type/format "should" be a bit easier to implement.

**Comment by Banthaf0dder** [comment]
> I'm going to re-open this issue. It's not high-priority (unless we see a lot of 👍) since we have the option in PowerShell, but I can see value in what is being asked for.
> 
> We still have more work to completely move everything the CLI is calling into the COM APIs, and once that's done, providing some kind of output type/format "should" be a bit easier to implement.

I'd like to note one thing in reference to powershell. Until the winget powershell module can be used in system context there is a gap in functionality that could be filled if winget output could be json formatted. That would make converting winget output to a powershell object very easy.

**Comment by sam-bryant** [comment]
Can we get an update on this?  It would be a very beneficial feature to output structured data like most other modern cli's.

**Comment by denelon** [comment]
Hey everyone, we've been publishing the Microsoft.WinGet.Client PowerShell module for the last several months with each WinGet release. It now works on Windows PowerShell and it works in system context (depending on the installer one is trying to use, and the scope is "System").

**Comment by svrooij** [comment]
> Hey everyone, we've been publishing the Microsoft.WinGet.Client PowerShell module for the last several months with each WinGet release. It now works on Windows PowerShell and it works in system context (depending on the installer one is trying to use, and the scope is "System").

This sounds nice, but is it is not installed with WinGet and it is `19MB`. I still feel that just outputting everything as json would still be a better solution.

**Comment by jo-chemla** [comment]
This thread shows there is some strong interest in JSON ready output for winget CLI commands like `list, show` etc - or probably any native machine-readable format - the raw package manifests yaml format included. 

For anyone interested in the powershell workaround, [here](https://powershellisfun.com/2024/11/28/using-the-powershell-winget-module/) is a link to a example documentation - execute the below in powershell admin terminal to install the module and play with it. 

```powershell
# not working: winget install Microsoft.WinGet.Client
Install-Module -Name Microsoft.WinGet.Client
Import-Module -Name Microsoft.WinGet.Client

Get-WingetPackage | convertto-JSON | Out-File -filepath "wingetPackage-list.json"
$ids = (cat wingetPackage-list.json | jq '.[] | select(.Source == \"winget\") | .Id')
```

Unfortunately, there do not seem to be an equivalent PowerShell cmdlet to the `winget show` CLI which would dump `Installer Url, moniker, License` etc, so these attributes would have to be parsed from raw console output via eg the below snippet

```powershell
$show_result = winget show $id
$urls = @(
    $show_result |
        Select-String -Pattern 'Installer Url:\s*(.*)' |
        ForEach-Object { $_.Matches.Groups[1].Value.Trim() }
)

```
